tests/backends/hardware/sr/v4/test_power_board.py

```diff
@@ -0,0 +1,470 @@
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    assert backend.firmware_version == 3
+
+
+def test_backend_bad_firmware_version():
+    """Test that we can get the firmware version."""
+    device = MockUSBPowerBoardDevice("SERIAL0", fw_version=2)
+    with pytest.raises(NotImplementedError):
+        SRV4PowerBoardHardwareBackend(device)
+
+
+def test_backend_serial_number():
+    """Test that we can get the serial number."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    assert backend.serial() == "SERIAL0"
+
+
+def test_backend_get_power_output_enabled():
+    """Test that we can read the enable status of a PowerOutput."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    for i in range(0, 6):
+        assert not backend.get_power_output_enabled(MockBoard(), i)
+
+    with pytest.raises(KeyError):
+        backend.get_power_output_enabled(MockBoard(), 6)
+
+
+def test_backend_set_power_output_enabled():
+    """Test that we can read the enable status of a PowerOutput."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    for i in range(0, 6):
+        backend.set_power_output_enabled(MockBoard(), i, True)
+
+    with pytest.raises(ValueError):
+        backend.set_power_output_enabled(MockBoard(), 6, True)
+
+
+def test_backend_get_power_output_current():
+    """Test that we can read the current on a PowerOutput."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    for i in range(0, 6):
+        assert 1.2 == backend.get_power_output_current(MockBoard(), i)
+
+    with pytest.raises(ValueError):
+        backend.get_power_output_current(MockBoard(), 6)
+
+
+def test_backend_piezo_buzz():
+    """Test that we can buzz the Piezo."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    # Buzz a Note
+    backend.buzz(MockBoard(), 0, timedelta(seconds=10), Note.D7)
+
+    # Buzz a frequency
+    backend.buzz(MockBoard(), 0, timedelta(seconds=10), 100)
+
+    # Buzz for too long.
+    with pytest.raises(ValueError):
+        backend.buzz(MockBoard(), 0, timedelta(seconds=100), 10)
+
+    # Test non-existent buzzer
+    with pytest.raises(ValueError):
+        backend.buzz(MockBoard(), 1, timedelta(seconds=10), 0)
+
+
+def test_backend_get_button_state():
+    """Test that we can get the button state."""
+    device = MockUSBPowerBoardDevice("SERIAL0")
+    backend = SRV4PowerBoardHardwareBackend(device)
+
+    assert not backend.get_button_state(MockBoard(), 0)
+
+    with pytest.raises(ValueError):
+        backend.get_button_state(MockBoard(), 1)
+
+
+def test_backend_wait_until_button_pressed():
+    """Test the wait_until_button_pressed."""
+    # No tests here as no known practical method to test it.
```

Comment:
Maybe we could inject `time.sleep` in the same manner we inject `usb.core.find`, and implement the test like so:

```python
device = MockUSBPowerBoardDevice("SERIAL0")

N = 3
def mock_sleep(secs):
  assert secs <= 0.1 # Check we poll the button at an acceptable rate.
  mock_sleep.times_called += 1
  # Button gets pressed during the N'th call to mock_sleep.
  if mock_sleep.times_called == N:
    device.button_state = True
  # Check mock_sleep doesn't get called more than N times. Prevents wait_until_button_pressed from looping forever.
  assert mock_sleep.times_called <= N
mock_sleep.times_called = 0

backend = SRV4PowerBoardHardwareBackend(device)
backend.wait_until_button_pressed(MockBoard(), 0, sleep=mock_sleep)
assert mock_sleep.times_called == N
```

If the above isn't feasible, please remove this empty test function since it is redundant.

Reply:
I attempted this, but it was very dirty. I'm going to accept the lack of coverage.

Reply:
Yeah, it was a rough idea - I had no idea how well it would pan out in practice.

And yeah, I would prefer us not having coverage of function to us having "fake" coverage (tests that don't verify anything).